File: apps/android/core/sync/src/main/kotlin/com/sigmundgranaas/turbo/expressive/core/sync/SharingRepository.kt
package com.sigmundgranaas.turbo.expressive.core.sync

import com.sigmundgranaas.turbo.expressive.core.auth.AuthConfig
import com.sigmundgranaas.turbo.expressive.core.common.Outcome
import io.ktor.client.call.body
import io.ktor.client.request.parameter
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import javax.inject.Inject

/** A redeemed share link: which resource the current user just gained access to, and the role. */
data class LinkRedemption(val resourceId: String, val resourceType: String, val role: String)

/**
 * Authenticated access to the backend Sharing service. Reuses the sync layer's
 * [AuthorizedHttp] (token + refresh-on-401). Web base for shareable URLs differs
 * from the API base.
 */
interface SharingRepository {
    /** The user's shareable friend code (e.g. "turbo-AB12CD"). */
    suspend fun friendCode(): Outcome<String>

    /** Create a share link for a resource; returns a shareable URL containing the link token. */
    suspend fun createLink(resourceId: String, role: String = ROLE_VIEWER): Outcome<String>

    /** Redeem a share-link token, granting the current user access to the resource. */
    suspend fun redeemLink(token: String): Outcome<LinkRedemption>

    /** Delta of resources shared with the current user (since the last cursor). */
    suspend fun sharedResources(since: String?): Outcome<ResourceSyncPageDto>

    // ── Friends + groups (default to "unsupported" so test fakes need not override) ──
    suspend fun friendships(): Outcome<List<FriendshipDto>> = unsupported()
    suspend fun requestFriendship(otherUserId: String): Outcome<Unit> = unsupported()
    suspend fun acceptFriendship(otherUserId: String): Outcome<Unit> = unsupported()
    suspend fun removeFriendship(otherUserId: String): Outcome<Unit> = unsupported()
    /** Resolve a friend code (with or without the "turbo-" prefix) to a user id. */
    suspend fun lookupUser(code: String): Outcome<String> = unsupported()
    suspend fun groups(): Outcome<List<GroupDto>> = unsupported()
    suspend fun createGroup(name: String): Outcome<GroupDto> = unsupported()
    suspend fun addGroupMember(groupId: String, userId: String): Outcome<Unit> = unsupported()
    suspend fun removeGroupMember(groupId: String, userId: String): Outcome<Unit> = unsupported()

    companion object {
        const val ROLE_VIEWER = "viewer"
        private fun <T> unsupported(): Outcome<T> = Outcome.Failure(UnsupportedOperationException())
    }
}

class KtorSharingRepository @Inject constructor(
    private val http: AuthorizedHttp,
) : SharingRepository {
    private val base = "${AuthConfig.BASE_URL}/api/sharing"

    override suspend fun friendCode(): Outcome<String> = runCatching {
        val resp = http.request("$base/me/profile") { method = HttpMethod.Get }
        check(resp.status.isSuccess()) { "profile ${resp.status}" }
        val dto: UserProfileDto = resp.body()
        "$FRIEND_CODE_PREFIX${dto.friendCode}"
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun createLink(resourceId: String, role: String): Outcome<String> = runCatching {
        val resp = http.request("$base/grants/links") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(GrantAsLinkRequest(resourceId = resourceId, role = role))
        }
        check(resp.status.isSuccess()) { "createLink ${resp.status}" }
        val dto: LinkGrantDto = resp.body()
        "$WEB_BASE_URL/link/${dto.linkToken}"
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun redeemLink(token: String): Outcome<LinkRedemption> = runCatching {
        val resp = http.request("$base/grants/links/$token/redeem") { method = HttpMethod.Post }
        check(resp.status.isSuccess()) { "redeem ${resp.status}" }
        val dto: LinkRedemptionDto = resp.body()
        LinkRedemption(dto.resourceId, dto.resourceType, dto.role)
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun sharedResources(since: String?): Outcome<ResourceSyncPageDto> = runCatching {
        val resp = http.request("$base/resources/sync") {
            method = HttpMethod.Get
            if (!since.isNullOrBlank()) parameter("since", since)
        }
        check(resp.status.isSuccess()) { "resources/sync ${resp.status}" }
        resp.body<ResourceSyncPageDto>()
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun friendships(): Outcome<List<FriendshipDto>> = runCatching {
        val resp = http.request("$base/friendships") { method = HttpMethod.Get }
        check(resp.status.isSuccess()) { "friendships ${resp.status}" }
        resp.body<List<FriendshipDto>>()
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun requestFriendship(otherUserId: String): Outcome<Unit> =
        friendshipAction("request", otherUserId)

    override suspend fun acceptFriendship(otherUserId: String): Outcome<Unit> =
        friendshipAction("accept", otherUserId)

    private suspend fun friendshipAction(action: String, otherUserId: String): Outcome<Unit> = runCatching {
        val resp = http.request("$base/friendships/$action") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(FriendshipActionRequest(otherUserId))
        }
        check(resp.status.isSuccess()) { "friendships/$action ${resp.status}" }
        Unit
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun removeFriendship(otherUserId: String): Outcome<Unit> = runCatching {
        val resp = http.request("$base/friendships/$otherUserId") { method = HttpMethod.Delete }
        check(resp.status.isSuccess()) { "friendships delete ${resp.status}" }
        Unit
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun lookupUser(code: String): Outcome<String> = runCatching {
        val bare = code.trim().removePrefix(FRIEND_CODE_PREFIX)
        val resp = http.request("$base/users/lookup") {
            method = HttpMethod.Get
            parameter("code", bare)
        }
        check(resp.status.isSuccess()) { "users/lookup ${resp.status}" }
        resp.body<UserLookupResponse>().userId
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun groups(): Outcome<List<GroupDto>> = runCatching {
        val resp = http.request("$base/groups") { method = HttpMethod.Get }
        check(resp.status.isSuccess()) { "groups ${resp.status}" }
        resp.body<List<GroupDto>>()
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun createGroup(name: String): Outcome<GroupDto> = runCatching {
        val resp = http.request("$base/groups") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(CreateGroupRequest(name))
        }
        check(resp.status.isSuccess()) { "createGroup ${resp.status}" }
        resp.body<GroupDto>()
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun addGroupMember(groupId: String, userId: String): Outcome<Unit> = runCatching {
        val resp = http.request("$base/groups/$groupId/members") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(GroupMemberRequest(userId))
        }
        check(resp.status.isSuccess()) { "addGroupMember ${resp.status}" }
        Unit
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    override suspend fun removeGroupMember(groupId: String, userId: String): Outcome<Unit> = runCatching {
        val resp = http.request("$base/groups/$groupId/members/$userId") { method = HttpMethod.Delete }
        check(resp.status.isSuccess()) { "removeGroupMember ${resp.status}" }
        Unit
    }.fold({ Outcome.Success(it) }, { Outcome.Failure(it) })

    private companion object {
        const val FRIEND_CODE_PREFIX = "turbo-"
        // Where shared links resolve (the web app / App Link host). Provisional path until P4 deep-link wiring.
        const val WEB_BASE_URL = "https://kart.sandring.no"
    }
}

File: apps/ios/Sources/CoreSync/SharingRepository.swift
import Foundation
import CoreCommon

/// A redeemed share link: which resource the user gained access to, and the role.
/// Mirrors `core.sync.LinkRedemption`.
public struct LinkRedemption: Equatable, Sendable {
    public let resourceId: String
    public let resourceType: String
    public let role: String
    public init(resourceId: String, resourceType: String, role: String) {
        self.resourceId = resourceId
        self.resourceType = resourceType
        self.role = role
    }
}

/// Authenticated access to the backend sharing service — friend code + share
/// links. Only usable when signed in (token-authed); the app hides sharing
/// otherwise. Mirrors `core.sync.SharingRepository`.
public protocol SharingRepository: Sendable {
    /// The user's shareable friend code (e.g. "turbo-AB12CD").
    func friendCode() async -> Outcome<String>
    /// Create a share link for a resource; returns a shareable URL.
    func createLink(resourceId: String, role: String) async -> Outcome<URL>
    /// Redeem a share-link token, granting the current user access.
    func redeemLink(token: String) async -> Outcome<LinkRedemption>
}

public extension SharingRepository {
    func createLink(resourceId: String) async -> Outcome<URL> {
        await createLink(resourceId: resourceId, role: "viewer")
    }
}

/// HTTP sharing client against the API (token-authed). Network path — the pure
/// pieces (`friendCode` formatting, link URL, redemption decode) are static + tested.
public struct HttpSharingRepository: SharingRepository {
    static let friendCodePrefix = "turbo-"

    private let apiBaseURL: URL
    private let webBaseURL: URL
    private let token: @Sendable () async -> String?
    private let session: URLSession

    public init(apiBaseURL: URL, webBaseURL: URL, token: @escaping @Sendable () async -> String?, session: URLSession = .shared) {
        self.apiBaseURL = apiBaseURL
        self.webBaseURL = webBaseURL
        self.token = token
        self.session = session
    }

    public func friendCode() async -> Outcome<String> {
        await request("sharing/me/profile", method: "GET") { data in
            guard let dto = try? JSONDecoder().decode(ProfileDTO.self, from: data) else { return nil }
            return Self.formatFriendCode(dto.friendCode)
        }
    }

    public func createLink(resourceId: String, role: String) async -> Outcome<URL> {
        let body = try? JSONSerialization.data(withJSONObject: ["resourceId": resourceId, "role": role])
        return await request("sharing/grants/links", method: "POST", body: body) { data in
            guard let dto = try? JSONDecoder().decode(LinkGrantDTO.self, from: data) else { return nil }
            return Self.linkURL(webBase: webBaseURL, token: dto.linkToken)
        }
    }

    public func redeemLink(token: String) async -> Outcome<LinkRedemption> {
        await request("sharing/grants/links/\(token)/redeem", method: "POST") { data in
            Self.decodeRedemption(data)
        }
    }

    // MARK: - Pure helpers (tested)

    static func formatFriendCode(_ code: String) -> String { "\(friendCodePrefix)\(code)" }

    static func linkURL(webBase: URL, token: String) -> URL {
        webBase.appendingPathComponent("link").appendingPathComponent(token)
    }

    static func decodeRedemption(_ data: Data) -> LinkRedemption? {
        guard let dto = try? JSONDecoder().decode(RedemptionDTO.self, from: data) else { return nil }
        return LinkRedemption(resourceId: dto.resourceId, resourceType: dto.resourceType, role: dto.role)
    }

    // MARK: - HTTP

    private func request<T>(_ path: String, method: String, body: Data? = nil,
                            decode: @Sendable (Data) -> T?) async -> Outcome<T> {
        var req = URLRequest(url: apiBaseURL.appendingPathComponent(path))
        req.httpMethod = method
        if let body { req.httpBody = body; req.setValue("application/json", forHTTPHeaderField: "Content-Type") }
        if let token = await token() { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode),
                  let value = decode(data) else {
                return .failure(NSError(domain: "Turbo.Sharing", code: 1))
            }
            return .success(value)
        } catch {
            return .failure(error)
        }
    }

    private struct ProfileDTO: Decodable { let friendCode: String }
    private struct LinkGrantDTO: Decodable { let linkToken: String }
    private struct RedemptionDTO: Decodable { let resourceId: String; let resourceType: String; let role: String }
}

/// In-memory sharing double for tests / hermetic UI runs.
public actor InMemorySharingRepository: SharingRepository {
    private let code: String
    private var links: [String: String] = [:]   // token → resourceId

    public init(code: String = "turbo-TEST01") { self.code = code }

    public func friendCode() -> Outcome<String> { .success(code) }

    public func createLink(resourceId: String, role: String) -> Outcome<URL> {
        let token = UUID().uuidString.prefix(8).lowercased()
        links[String(token)] = resourceId
        return .success(URL(string: "https://kart.sandring.no/link/\(token)")!)
    }

    public func redeemLink(token: String) -> Outcome<LinkRedemption> {
        guard let resourceId = links[token] else { return .failure(NSError(domain: "Turbo.Sharing", code: 404)) }
        return .success(LinkRedemption(resourceId: resourceId, resourceType: "marker", role: "viewer"))
    }
}
